### ISSUE TITLE
`file_event_callback:  EPERM` error notification after some background filesystem actions

### ISSUE THREAD

**Issue by yehorb** [nvim-neo-tree/neo-tree.nvim#914]
Windows 10, neovim 0.9.0, neo-tree 2.59.0.

In some cases, doing filesystem actions - such as `git checkout`, deleting directories using a shell command, using explorer, or even deleting nested directories using neo-tree itself, - causes neo-tree to soft-lock. The `nvim` becomes unresponsive and an endless stream of notifications with the text `file_event_callback:  EPERM` appears. There is no way to break out of the soft-lock other than killing the `nvim` process.

![Screenshot 2023-05-05 131335](https://user-images.githubusercontent.com/17297726/236432086-e172dd79-d0a6-4ee5-b391-31c136595a07.png)

Sadly, the issue is not easily reproducible, and I can not provide exact steps to get it. Sometimes `neo-tree` handles background file changes, and sometimes it breaks.

**Comment by WillEhrendreich** [comment]
This is a huge issue for me too, I'm glad I'm not the only one! Locks up to the point where have to quit neovim and restart. Seems unpredictable one way or the next. The little bit I've been able to track down about it seems to point to something involving node? I'm not sure. I hope someone has some insight, because I'd love it if there was no issue.

**Comment by aslaii** [comment]
Even after that error, the vim breaks meaning I had to close the whole terminal and open it again.

**Comment by bartektelec** [comment]
This happens to me pretty much every time I delete a directory using neotree

**Comment by nhat-vo** [comment]
@bartektelec Can you provide additional information like OS, neovim version, etc. and perhaps the exact steps you took to reproduce this please? I have tried for a while but I could not reproduce this error.

**Comment by bartektelec** [comment]
> @bartektelec Can you provide additional information like OS, neovim version, etc. and perhaps the exact steps you took to reproduce this please? I have tried for a while but I could not reproduce this error.

Sure,
Windows 11 Pro 22H2
NVIM v0.9.1
Most recent Neo-tree installed through Lazy
Using LazyVIM distro

What I do:
1. Open up Neovim in any kind of directory, but for this example let it be an empty dir
2. I go <leader>e to open up Neo-tree
3. I create two folders pressing `a` in Neo-tree - called A and B
4. I can delete them just fine at any moment, but if I ever unfold one of them in Neo-tree and then delete it the problem occurs
5. This can also happen when I press `a` on a folded directory, it automatically unfolds, and when I fold it and delete it later it also crashes
6. If I create a dir in neo-tree and then create a file in it using windows explorer so the neo-tree doesn't unfold then I can delete the directory just fine

When deleting an unfolded / opened directory:
![neotreeerror](https://github.com/nvim-neo-tree/neo-tree.nvim/assets/20260038/f21bd259-865f-4e3b-b140-559c55795e7e)

When I open, then close then delete:
![neotreeerror2](https://github.com/nvim-neo-tree/neo-tree.nvim/assets/20260038/eb1b99aa-8d8d-4617-a6be-3af3daf31129)

**Comment by yehorb** [comment]
@bartektelec, thank you for the reproduction.

Can confirm the same issue.

Windows 10 Pro 22H2.
NVIM v0.9.1.
neo-tree.nvim 2.70.0.

Steps to reproduce in my case:

1. Create an empty directory.
2. `cd` into it.
3. `nvim .` so that `neo-tree` is the buffer I see.
4. Create a directory `a` using `neo-tree`.
5. Create a directory `b` using `neo-tree`.
6. Create an empty file `foo.txt` in `a` directory (so it "unfolds") and write it to disk.
7. Delete `a` directory using `neo-tree` OR using `!rm -Recurse -Force a`.
8. Bug reproduced.

Looks like this behavior is caused by deleting "unfolded" nodes, either via `neo-tree` or by other means.

Is reproducible on Windows.

**Comment by nhat-vo** [comment]
I have tried on my computer with:
```
Windows Education 22H2 (OS Build 22621.1992)
nvim v0.9.1
neo-tree 3.0.0 (commit 981a43a)
```
and the following configuration file:
```
vim.g.mapleader = " "
vim.g.maplocalleader = " "

local lazypath = vim.fn.stdpath("data") .. "/lazy/lazy.nvim"
if not vim.loop.fs_stat(lazypath) then
  vim.fn.system({
    "git",
    "clone",
    "--filter=blob:none",
    "https://github.com/folke/lazy.nvim.git",
    "--branch=stable", -- latest stable release
    lazypath,
  })
end
vim.opt.rtp:prepend(lazypath)

require("lazy").setup({
  {
    "nvim-neo-tree/neo-tree.nvim",
    version = "*",
    dependencies = {
      "nvim-lua/plenary.nvim",
      "nvim-tree/nvim-web-devicons",
      "MunifTanjim/nui.nvim",
    },
    config = function()
      require("neo-tree").setup({
        close_if_last_window = true,
      })
    end,
  },
})
```

but the issue does not seem to appear :(. Could you have a look to see if the issue still persists with this configuration, please? @yehorb @bartektelec 

https://github.com/nvim-neo-tree/neo-tree.nvim/assets/30724347/6700433b-9eb2-4f09-943b-14c0e8af52c1

**Comment by bartektelec** [comment]
Can confirm that overriding the `config` property with the one you provided fixes the issue.
That's because it overrides the whole LazyVIM's opts record.

I narrowed the issue down to be one of the defaults LazyVIM provides:
```lua
        use_libuv_file_watcher = true,
```

This is what causes neovim to crash when deleting a directory, however setting it to false will cause neo-tree not to refresh when I delete the directory using a different file explorer so this can be seen as a temporary fix. 

Not sure why libuv would cause it to fail

**Comment by nhat-vo** [comment]
@bartektelec Thanks for your insights! I'm not sure why using `libuv` causes this behavior - it is probably some bug between libuv and windows path. I will have a look to see if I could identify the source of the problem, but for now we can turn of the `libuv` as a temporary fix like you said.

**Comment by samuelnihoul** [comment]
I think the issue is linked to Windows itself. It's always pretty hard to delete files in there. Especially if you'r trying to delete a non-empty folder, or if a file you're trying to delete is open, etc.

**Comment by samuelnihoul** [comment]
> @bartektelec Can you provide additional information like OS, neovim version, etc. and perhaps the exact steps you took to reproduce this please? I have tried for a while but I could not reproduce this error.

Try to delete a non-empty folder with a file that is open in the editor inside of it.

**Comment by WillEhrendreich** [comment]
I don't think this issue should be counted as closed, per se, as the solution of "just don't use the feature on windows" is kind of a bummer, and this person in the issue mentioned seems to think it's a neo-tree issue. https://github.com/libuv/libuv/issues/3267#issuecomment-1804273269

**Comment by cseickel** [comment]
> I don't think this issue should be counted as closed, per se, as the solution of "just don't use the feature on windows" is kind of a bummer,

Agreed. I'll re-open, but I'm still waiting on a Windows user to debug and submit a fix.

**Comment by nishnat-rishi** [comment]
I can't open a fix for this because I don't know enough about any of this, but I have a solution that Claude worked out for me. This fixes both the error spam and implements proper debouncing of filesystem events. The AI told me to relay that the key changes:

- Handle Windows EPERM errors as expected events rather than errors
- Batch filesystem events and refresh at common parent directories
- Add appropriate delays to let Windows filesystem operations complete

Here's parts of the `fs_scan.lua` including all the modifications that resolved the issue:

```lua
-- This files holds code for scanning the filesystem to build the tree.
local uv = vim.loop

local renderer = require("neo-tree.ui.renderer")
local utils = require("neo-tree.utils")
local filter_external = require("neo-tree.sources.filesystem.lib.filter_external")
local file_items = require("neo-tree.sources.common.file-items")
local file_nesting = require("neo-tree.sources.common.file-nesting")
local log = require("neo-tree.log")
local fs_watch = require("neo-tree.sources.filesystem.lib.fs_watch")
local git = require("neo-tree.git")
local events = require("neo-tree.events")
local async = require("plenary.async")

local M = {}

local is_refreshing = {}
local recent_errors = {}
local pending_refresh = nil

local get_common_parent = function(path1, path2)
  if not path1 or not path2 then
    return path1 or path2
  end

  local parts1 = vim.split(path1, utils.path_separator)
  local parts2 = vim.split(path2, utils.path_separator)

  local common = {}
  for i = 1, math.min(#parts1, #parts2) do
    if parts1[i] == parts2[i] then
      table.insert(common, parts1[i])
    else
      break
    end
  end

  return table.concat(common, utils.path_separator)
end

local on_directory_loaded = function(context, dir_path)
  local state = context.state
  local scanned_folder = context.folders[dir_path]
  if scanned_folder then
    scanned_folder.loaded = true
  end
  if state.use_libuv_file_watcher then
    local root = context.folders[dir_path]
    if root then
      local target_path = root.is_link and root.link_to or root.path
      local fs_watch_callback = vim.schedule_wrap(function(err, fname)
        if err then
          if vim.startswith(err, "EPERM") or vim.startswith(err, "EACCES") then
            local now = vim.loop.now()
            recent_errors[target_path] = now

            -- If we already have a pending refresh, find the common parent
            if pending_refresh then
              pending_refresh.path = get_common_parent(pending_refresh.path, target_path)
              pending_refresh.last_update = now
              return
            end

            -- Create new pending refresh
            pending_refresh = {
              path = target_path,
              last_update = now,
              timer = vim.defer_fn(function()
                if pending_refresh then
                  local debug_file = io.open(vim.fn.stdpath("data") .. "/neo-tree-debug.log", "a")
                  debug_file:write(
                    string.format(
                      "[DEBUG %s] Executing refresh for path %s\n",
                      os.date(),
                      pending_refresh.path
                    )
                  )
                  debug_file:flush()
                  debug_file:close()

                  events.fire_event(events.FS_EVENT, { afile = pending_refresh.path })
                  pending_refresh = nil

                  -- Clean up old error records
                  for path, time in pairs(recent_errors) do
                    if (now - time) > 1000 then
                      recent_errors[path] = nil
                    end
                  end
                end
              end, 300),
            }
          else
            log.error("file_event_callback: ", err)
          end
          return
        end

        if context.is_a_never_show_file(fname) then
          -- don't fire events for nodes that are designated as "never show"
          return
        else
          events.fire_event(events.FS_EVENT, { afile = target_path })
        end
      end)

      log.trace("Adding fs watcher for ", target_path)
      fs_watch.watch_folder(target_path, fs_watch_callback)
    end
  end
end
... -- rest would be unchanged
```

Someone with actual skill please utilize this (if it makes sense) and submit a PR. Thanks!
